libp2p-floodsub/src/lib.rs

```diff
@@ -0,0 +1,640 @@
+                                    }
+                                    Err(err) => future::err(err),
+                                };
+                                Box::new(fut) as Box<_>
+                            }
+
+                            Some((bytes, true)) => {
+                                // Received a packet from the channel.
+                                // Need to send a message to remote.
+                                trace!(target: "libp2p-floodsub", "Effectively sending message \
+                                                               to remote");
+                                let future = floodsub_sink.send(bytes).map(|floodsub_sink| {
+                                    future::Loop::Continue((floodsub_sink, rest))
+                                });
+                                Box::new(future) as Box<_>
+                            }
+
+                            None => {
+                                // Both the connection stream and `rx` are empty, so we break
+                                // the loop.
+                                trace!(target: "libp2p-floodsub", "Pubsub future clean finish");
+                                // TODO: what if multiple connections?
+                                inner.remote_connections.write().remove(&remote_addr);
+                                let future = future::ok(future::Loop::Break(()));
+                                Box::new(future) as Box<Future<Item = _, Error = _>>
+                            }
+                        }
+                    })
+            },
+        );
+
+        future::ok(FloodSubFuture {
+            inner: Box::new(future) as Box<_>,
+        })
+    }
+}
+
+/// Allows one to control the behaviour of the floodsub system.
+#[derive(Clone)]
+pub struct FloodSubController<T, C>
+where
+    T: MuxedTransport + 'static,                // TODO: 'static :-/
+    C: ConnectionUpgrade<T::RawConn> + 'static, // TODO: 'static :-/
+{
+    inner: Arc<Inner>,
+    swarm: SwarmController<T, C>,
+}
+
+struct Inner {
+    // Our local peer ID multihash, to pass as the source.
+    peer_id: Vec<u8>,
+
+    // Channel where to send the messages that should be dispatched to the user.
+    output_tx: mpsc::UnboundedSender<Message>,
+
+    // Active connections with a remote.
+    remote_connections: RwLock<FnvHashMap<Multiaddr, RemoteInfo>>,
+
+    // List of topics we're subscribed to. Necessary in order to filter out messages that we
+    // erroneously receive.
+    subscribed_topics: RwLock<Vec<Topic>>,
+
+    // Sequence number for the messages we send.
+    seq_no: AtomicUsize,
+
+    // We keep track of the messages we received (in the format `(remote ID, seq_no)`) so that we
+    // don't dispatch the same message twice if we receive it twice on the network.
+    // TODO: the `HashSet` will keep growing indefinitely :-/
+    received: Mutex<FnvHashSet<(Vec<u8>, Vec<u8>)>>,
```

Comment:
Two options to bound memory usage:
1) Make this a bloom filter, which would also skip 2(!) vector allocations per element but could devolve if every node gets its filters saturated.
2) Make this a FIFO, which would mean slower `get` performance but you probably don't need to keep more than, say, the last couple of hundred messages.

Also, to reduce memory usage you could have a `FnvHashSet<u64>` and hash the elements, since Rust will store the keys but we don't need them. The memory usage will still be unbounded but at least each element will only take up 8 bytes

Reply:
I'm not sure how problematic it is to send the same message multiple times, which could happen with a bloom filter or a FIFO.